Add alias dispatch key Math.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44993 Define macro for DispatchKey::toString.
* #44933 Align casing in test_dispatch with dispatch keys.
* #44556 Support Math keyword in native_functions.yaml.
* **#44354 Add alias dispatch key Math.**

Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

Commit:
Add alias dispatch key Math.

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."




Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -906,6 +972,65 @@ TEST(OperatorRegistrationTest, AutogradBackendOverridesAutogradKernel) {
   EXPECT_FALSE(called_nonautograd);
 }
 
+TEST(OperatorRegistrationTest, AutogradBackendOverridesMathKernel) {
```
If I'm right about the need to avoid choosing Autograd when no backend kernel is registered, we should probably add (or replace this with) a test for that.

bhosmer commented on c10/core/DispatchKey.cpp:
```diff
@@ -96,6 +96,9 @@ const char* toString(DispatchKey t) {
       return "VmapMode";
 
+    case DispatchKey::Math:
+      return "Math";
```
I don't know if you wanna do it in this PR, but it might be time to start using macros [a la ScalarType.h](https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h#L25-L41) to define these sorts of functions...

ailzhang replied:
tracked in https://github.com/pytorch/pytorch/issues/44657 :D

bhosmer commented on c10/core/DispatchKey.h:
```diff
@@ -331,7 +332,7 @@ constexpr DispatchKey kAutograd = DispatchKey::Autograd;
 // Check if a DispatchKey is an alias mapping to other runtime keys.
 inline bool isAliasDispatchKey(DispatchKey k) {
-  return k == DispatchKey::Autograd;
+  return k == DispatchKey::Autograd || k == DispatchKey::Math;
```
How about `k > NumDispatchKeys && k <= EndOfAliasKeys`? Then it'll track additions automatically

bhosmer commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -12,10 +12,46 @@ constexpr DispatchKeySet autograd_dispatch_keyset = DispatchKeySet({
   DispatchKey::AutogradOther,
 });
 
+constexpr DispatchKeySet math_dispatch_keyset = DispatchKeySet({
```
We shouldn't need to duplicate these big slabs - ideally there'd be a single source of truth for each alias keyset. I'd say let's build the `DispatchKeySet`s from the `std::array`s defined further down, and have the arrays be the source of truth. (If there's some reason that doesn't work straightforwardly, maybe that's another reason to go to macros...)

ailzhang replied:
actually the iterator for DispatchKeySet landed so I'll rebase and remove the std::array one ;)

bhosmer replied:
😍

bhosmer commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -12,10 +12,46 @@ constexpr DispatchKeySet autograd_dispatch_keyset = DispatchKeySet({
   switch (t) {
     case DispatchKey::Autograd:
       return autograd_dispatch_keyset;
+    case DispatchKey::Math:
```
Same DRY note for this function - we shouldn't be repeating this mapping if we can help it, instead ideally `getRuntimeDispatchKeySet(t)` would be based on `getRuntimeDispatchKeys(t)` which "owns" the mapping.

bhosmer commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
 AutogradCPU: fallthrough registered at /dev/null:0 [backend fallback]
 AutogradCUDA: default_def_name_t_t [catch all]
 AutogradXLA: default_def_name_t_t [catch all]
+''')
```
Haven't looked closely at these, but just tossing in a repeat of the comment about Autograd vs Math from above

bhosmer reviewed (commented):
Looking good! But I want to make sure I understand your thinking on the Math vs Autograd precedence, per inline comments :)

ailzhang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -160,18 +160,21 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
-  //      alias keys AutogradDispatchCPUOrCUDA and Math.
-  // TODO: we can fix (2.2) and remove (4) after TypeDefault registrations are moved from catchAll to Math
+  // Alias Key Precedence:
+  //   Math > Autograd
```
@bhosmer I think b) is also implicitly covered in the current logic where `Math>Autograd` + a)? 
When we have registration to both `Math` and `Autograd`:
1. If we don't have registration to backend X, for `AutogradX` it'll pick up `Math` kernel since `Math` has higher precedence. <= this is b? 
2. If we have registration to backend X, for `AutogradX` it'll skip 2.1 (due to `a`) and pick up `Autograd` kernel.

bhosmer replied:
Yeah I think that's right, in the case where we have both a Math kernel and an Autograd kernel. But the case I'm thinking of is the one where we have an Autograd kernel but no Math kernel (and no backend kernel) - in that case simple precedence says to plug in the Autograd kernel anyway, but I think we want to avoid that. 

(BTW I'm pretty sure at least one of our trial use cases has this Autograd-but-no-Math configuration - you can see how it's handled in the little working prototype.)

ailzhang replied:
We agreed offline when we don't have `Math` and backend `X` registered, `AutogradX` should pick up kernel from `Autograd` and error out when running inference. 
I've updated the colab notebook to match this behavior. :D

ezyang replied:
Good. Now put this in the comments ;)

ezyang commented:
> Add alias dispatch key Math

For external observers, it will be helpful to replicate the relevant parts from the design doc into the PR summary.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -160,18 +160,21 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
   //  (2) Handle runtime keys that have kernels available from alias keys
-  //    (2.1) Use kernel from DispatchKey::Autograd if available
-  //    (2.2) For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
+  //    (2.1) Use kernel from DispatchKey::Math if available
```
Aren't you going to talk about the backend special case here?

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -181,23 +184,42 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
+    if (kern_math != kernels_.end()) {
+      TORCH_INTERNAL_ASSERT(!kern_math->second.empty());
+      TORCH_INTERNAL_ASSERT(kern_math->second.front().kernel.isValid());
+      if (!isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
```
Hoist this conditional outside of the `kern_math` lookup, there is no dependence on `kern_math`.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -181,23 +184,42 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
     auto kern_autograd = kernels_.find(DispatchKey::Autograd);
     if (kern_autograd != kernels_.end()) {
       TORCH_INTERNAL_ASSERT(!kern_autograd->second.empty());
       TORCH_INTERNAL_ASSERT(kern_autograd->second.front().kernel.isValid());
```
Wondering if we should factor this "lookup and then retrieve second.front()" into a helper function of some sort, it's showing up a lot here.

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -30,6 +30,7 @@ c10::optional<c10::DispatchKey> parseDispatchKey(const std::string& k) {
     {"cpu", c10::DispatchKey::CPU},
     {"cuda", c10::DispatchKey::CUDA},
     {"xla", c10::DispatchKey::XLA},
+    {"math", c10::DispatchKey::Math},
```
Hmm. Maybe we should replicate the casing of these entries, would make the Python tests easier to read. No need to do here. Maybe you can ask a bootcamper to do it.

ezyang commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
+        global_m = C._dispatch_library("IMPL", "_", "autogradcpu")
+        result = self.commute("foo", [
+            # m.def("foo", [](const Tensor & x) { return x })
+            lambda m: m.def_name_t_t("foo"),
```
OK, remind me: in the glorious new future, what key does `m.def("foo", foo_impl);` get registered to? It's going to be Math, right? So this test isn't very future compatible, because what you have here is an implicit definition of math here, and then an explicit definition of math below. When you eventually cut over, this test should start failing as duplicate definitions of the same type will be non-commutative.

I think probably what you intended here to do is to only `def_name` with out the actual default definition. That would give you give you a computed table with only a math entry, which sounds like what you implied in your test name.

ezyang replied:
It looks like something similar happened for all the tests below.

ezyang commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
+AutogradCPU: fallthrough registered at /dev/null:0 [backend fallback]
+AutogradCUDA: fn2 [math kernel]
+AutogradXLA: fn2 [math kernel]
+Meta: default_def_name_t_t [catch all]
```
This is like, totes not what someone who registered a catch all wanted, right? :) This is incredibly unlikely to do the right thing. Even math kernel would be better!

ezyang commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
+        # computed dispatch table is too big, so we only check on a few entries we're interested in.
+        extracted_table = extract_dispatch_table_with_keys(
+            table,
+            ('CPU', 'CUDA', 'XLA', 'AutogradOther', 'AutogradCPU', 'AutogradCUDA', 'AutogradXLA', 'Meta'))
```
Consider defining this as a global somewhere instead of replicating it.

ezyang commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
+CUDA: fn2 [math kernel]
+XLA: fn2 [math kernel]
+AutogradOther: fn2 [math kernel]
+AutogradCPU: fallthrough registered at /dev/null:0 [backend fallback]
```
Huh, where did this fallthrough come from? I don't see any fallthroughs in this test. Is this coming from one of our app-wide registrations and you're munging this? (I must have missed that in post review). If so, it would probably be good to be a little more creative with the file names here so we can easily distinguish between "things that came from this test suite" and "things that came from the framework" (even if we don't want to report their exact line numbers, since those are prone to wobbling.)

ezyang replied:
And I guess, in the terminal state, this is going to be a backend fallback saying that backwards is not supported, right? Not just a fallthrough.

ailzhang replied:
Yup this won't be a fallthrough in the terminal state.

ezyang commented on test/test_dispatch.py:
```diff
@@ -374,6 +374,114 @@ def test_computed_table_with_cpu_catchall(self):
+            # m.impl("foo", torch::kAutograd, [](const Tensor & x) { return x })
+            lambda m: m.impl_t_t("foo", "autograd", debug="fn2"),
+            # m.impl("foo", torch::kMath, [](const Tensor & x) { return x })
+            lambda m: m.impl_t_t("foo", "math", debug="fn3"),
```
If you give more descriptive debug names, e.g., `fn_autograd`, that would make it easier to interpret the results below.

ezyang commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -38,11 +74,45 @@ constexpr std::array<DispatchKey, 7> autograd_dispatch_keys {
     DispatchKey::AutogradPrivateUse1, DispatchKey::AutogradPrivateUse2,
     DispatchKey::AutogradPrivateUse3, DispatchKey::AutogradOther};
 
+constexpr std::array<DispatchKey, 30> math_dispatch_keys {
```
Should be a comment here describing what a "math dispatch key" should be and when you should include something in here

ezyang replied:
I'm wondering if we should generate these arrays more incrementally. For example, we could have an array for backends, an array for autograd keys, and then you'd paste them together to get math.

ezyang replied:
This is dependent on dispatch key set union being constexpr but there isnt' any reason it shouldn't be.

ezyang commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -877,6 +877,72 @@ TEST(OperatorRegistrationTest, whenRegisteringAutogradKernelWithCatchAllKernel_t
 
+TEST(OperatorRegistrationTest, RegisterMathKernel) {
+  auto registrar = c10::RegisterOperators().op("_test::dummy(Tensor dummy) -> ()", c10::RegisterOperators::options()
+    .kernel<decltype(nonautograd_kernel), &nonautograd_kernel>(DispatchKey::Math));
```
These tests should just be written in the new API. But if you trust the computed dispatch key tables in Python, I think it is not too important to test these in C++, besides the very basics.

ezyang reviewed (approved):
I had a bunch of substantive comments but the core logic looks good so I'm approving.

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,49 +153,71 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
-  //    (2.1) Use kernel from DispatchKey::Autograd if available
-  //    (2.2) For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
+  //    (2.1) Use kernel from DispatchKey::Math if available. For autograd keys, we only use kernel from Math
+  //          when there's no direct registration to its corresponding backend key.
```
Hey, sorry I didn't realize this during the earlier discussion but don't we have the same issue here regarding keys like AutogradOther that represent multiple backends? If any of them register a kernel, then registering a Math kernel to (e.g.) AutogradOther will cause it to be skipped, right?

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,49 +153,71 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Math)
+      && hasKernelForDispatchKey(DispatchKey::Math)
+      && !(isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+           && hasKernelForDispatchKey(getBackendKeyFromAutograd(dispatch_key)))) {
```
The "backend that uses AutogradOther" case will be missed here, bc `getBackendKeyFromAutograd(AutogradOther)` returns undefined - I think this is a problem. 

Most immediately, since the data model says there's a 1:n relationship between autograd keys and backend keys, I think the correct API for moving along this mapping is `getBackendKeySetFromAutograd(k)` that returns a keyset of backends in all cases (though ofc it might be a singleton). (Apologies that I didn't get a chance to review the PR where this went in!)

Once we have that, then the question becomes how to generalize the "only use Math if no backend kernel has been registered" condition to sets of backends. Unfortunately I think any simple answer - say, anding the "no kernel" constraint across the set - reintroduces the problem of nonlocal effects of a registration change. 

Going to leave this comment here as-is before thinking more about it, just in case I've made a basic mistake and it's not really a problem :) otherwise, let's chat about it at tomorrow morning's thing.

ailzhang replied:
Great point! It's true that current implementation will always pick up `Math` kernel if exists. Maybe a workaround could be `make Autograd higher precedence than Math for AutogradOther`? 
Happy to followup on this tmr morning :D

ezyang replied:
There is no right answer here. If it truly is a show stopper, the right thing to do is to move the key in question out of AutogradOther. I don't mind either default, but Math is probably easier for these exotic custom backends (e.g., the Math version implies less work for a custom backend; if they want to do more work, then they can also ask us to split out their key).

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,49 +153,71 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  // 2.2. For autograd backend keys, use kernel from DispatchKey::Autograd if available
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+      && hasKernelForDispatchKey(DispatchKey::Autograd)) {
+    return {getKernelForDispatchKey(DispatchKey::Autograd), "autograd kernel"};
```
Kind of an aside, but I'm mildly averse to the "hasThing/getThing" idiom for these helpers bc they repeat the linear search. NBD in precomputation, and the code is definitely neater with the lookup factored... but in a perfect world I'd probably have the helper return an optional, just in case code cargo-culted from here winds up in a hotspot. 

Eh I dunno if it's worth tearing up the sidewalk. Really I'm still thinking about the AutogradOther thing 😁

ailzhang replied:
Looks like we need to return an optional reference here. Just to double check, do we prefer ` optional<std::reference_wrapper<T>>` or a raw pointer if we want to do that? :P

ezyang replied:
The benefit of reference wrapper is that it is guaranteed to be non-null. cc @smessmer

bhosmer reviewed (approved):
Accepting on the assumption I'm wrong about the possible problem mentioned inline... if it's a bona fide issue let's dig in tmrw 😁

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

ailzhang commented:
New changes since last review:
For an op if we already have `Math` kernel for it and also a direct registration to a backend mapping to `AutogradOther` (say `SparseCPU`), we eagerly put an error kernel at `AutogradOther`. And when training is called on autograd other backends, this error kernel will report error and remind backend extenders to request a separate autograd key from us. 
(note inference on all backends and training on CPU/CUDA/XLA/Private* will still work fine)
Also added a test for this behavior here `throwsWhenRegisterToBackendMapsToAutogradOther` .

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +153,89 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  bool found = false;
+  for (auto k : ks) {
+    auto it = kernels_.find(k);
+    found = found || it != kernels_.end();
```
Not going to short circuit?

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +153,89 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  auto math_registration = getKernelForDispatchKey(DispatchKey::Math);
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Math) && math_registration) {
+    if (dispatch_key == DispatchKey::AutogradOther && is_autograd_key_with_backend_kernel) {
+      return {missingKernel_, "missing"};
```
Weren't you going to replace this with a different kernel that has the nice error message explaining how to get the user out of this situation?

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +153,89 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  //      when there's no direct registration to its corresponding backend key.
+  //      For AutogradOther, we return missing kernel if there's registration to any of
+  //      its backends.
+  auto math_registration = getKernelForDispatchKey(DispatchKey::Math);
```
Small opt: do the `math_registration` lookup inside the if statement, as you can skip it entirely if the key is not in the alias. We should care a little about this as this code is somewhat hot in the static initialization phase.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +153,89 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
+  //      when there's no direct registration to its corresponding backend key.
+  //      For AutogradOther, we return missing kernel if there's registration to any of
+  //      its backends.
```
This comment would be improved if it said why we are doing this.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -271,6 +307,12 @@ void OperatorEntry::checkInvariants() const {
+    "You're registering kernel to a backend maps to AutogradOther trying to override Math kernel behavior. "
+    "If this is intended, please open an issue to request a proper Autograd dispatch key for your backend.",
+    dumpState()
+  );
```
This doesn't look right. In general you can't expect `checkInvariants` will be called. If you want a user facing message, you should report the error elsewhere; maybe in dispatch table recalcluation?

ailzhang replied:
Aha I'll create a new kernel as you suggested above and put the error message there ;)

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

Commit:
Update on "Add alias dispatch key Math."


Design proposal: https://github.com/pytorch/pytorch/issues/44680

Differential Revision: [D23591481](https://our.internmc.facebook.com/intern/diff/D23591481)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +156,91 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  //          See Note [Ambiguity in AutogradOther kernel] for more details.
+  //    (2.2) Use kernel from DispatchKey::Autograd if available
+  //    (2.3) Special logic to handle catchAll for Autograd keys
+  //          For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
```
Also I think it's worth noting in the comments that the implementation relies on the invariant that  `computeDispatchTableEntryWithDebug()` will be called first for backend keys, then for autograd keys (or more precisely, for a given backend, it'll be called for that backend's autograd key *after* the backend key). In fact it might be worth specifically referencing the logic at the bottom of `updateDispatchTable_` that we're relying on to preserve the invariant.

bhosmer commented on c10/core/DispatchKeySet.cpp:
```diff
@@ -2,13 +2,71 @@
+  }
+}
+
+DispatchKeySet getBackendKeySetFromAutograd(DispatchKey t) {
```
Could do the same macro-based definition thing for autograd keys as for dispatch keys as a whole (in fact, the macro for the full set could call sub-macros for backend and autograd, I think?)

Again though not worth stopping and doing right away, but if you're back in here for other things it might be worth it.

bhosmer reviewed (commented):
Awesome! Threw in a handful of ideas for followups (mostly comments) inline, definitely nothing urgent 😁

ailzhang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -153,51 +156,91 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
+  //          For AutogradOther, we eagerly return ambiguousAutogradOtherKernel_ if there's registration to any of
+  //          its backends and ask backend extender to request a decicated Autograd key for the backend.
+  //          See Note [Ambiguity in AutogradOther kernel] for more details.
+  //    (2.2) Use kernel from DispatchKey::Autograd if available
```
Oh I think this part is covered by `(2) Handle runtime keys that have kernels available from alias keys`? as headline? ;)

bhosmer replied:
Ah you're totally right, the info is already there :) personally I don't mind redundant information in descriptions of tricky-ish setups like this, but yeah this is perfectly fine as is.
